pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/DistinctTable.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.pinot.common.data.FieldSpec;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.common.utils.DataTable;
+import org.apache.pinot.core.common.datatable.DataTableBuilder;
+import org.apache.pinot.core.common.datatable.DataTableFactory;
+import org.apache.pinot.core.data.table.Key;
+
+
+/**
+ * This serves the following purposes:
+ *
+ * (1) Intermediate result object for Distinct aggregation function
+ * (2) The same object is serialized by the server inside the data table
+ * for sending the results to broker. Broker deserializes it.
+ */
+public class DistinctTable {
+  private static final double LOAD_FACTOR = 0.75;
+  private static final int MAX_INITIAL_CAPACITY = 64 * 1024;
+  private FieldSpec.DataType[] _columnTypes;
+  private String[] _columnNames;
+  private Set<Key> _table;
+
+  /**
+   * Add a row to hash table
+   * @param key multi-column key to add
+   */
+  public void addKey(final Key key) {
+    _table.add(key);
+  }
+
+  public DistinctTable(int limit) {
+    // TODO: see if 64k is the right max initial capacity to use
+    // if it turns out that users always use LIMIT N > 0.75 * 64k and
+    // there are indeed that many records, then there will be resizes.
+    // The current method of setting the initial capacity as
+    // min(64k, limit/loadFactor) will not require resizes for LIMIT N
+    // where N <= 48000
+    int initialCapacity = Math.min(MAX_INITIAL_CAPACITY, Math.abs(nextPowerOfTwo((int) (limit / LOAD_FACTOR))));
+    _table = new HashSet<>(initialCapacity);
+  }
+
+  /**
+   * DESERIALIZE: Broker side
+   * @param byteBuffer data to deserialize
+   * @throws IOException
+   */
+  public DistinctTable(ByteBuffer byteBuffer)
+      throws IOException {
+    DataTable dataTable = DataTableFactory.getDataTable(byteBuffer);
+    DataSchema dataSchema = dataTable.getDataSchema();
+    int numRows = dataTable.getNumberOfRows();
+    int numColumns = dataSchema.size();
+
+    _table = new HashSet<>();
+
+    // extract rows from the datatable
+    for (int rowIndex = 0; rowIndex < numRows; rowIndex++) {
+      Object[] columnValues = new Object[numColumns];
+      for (int colIndex = 0; colIndex < numColumns; colIndex++) {
+        DataSchema.ColumnDataType columnDataType = dataSchema.getColumnDataType(colIndex);
+        switch (columnDataType) {
+          case INT:
+            columnValues[colIndex] = dataTable.getInt(rowIndex, colIndex);
+            break;
+          case LONG:
+            columnValues[colIndex] = dataTable.getLong(rowIndex, colIndex);
+            break;
+          case FLOAT:
+            columnValues[colIndex] = dataTable.getFloat(rowIndex, colIndex);
+            break;
+          case DOUBLE:
+            columnValues[colIndex] = dataTable.getDouble(rowIndex, colIndex);
+            break;
+          case STRING:
+            columnValues[colIndex] = dataTable.getString(rowIndex, colIndex);
+            break;
+          default:
```

Comment:
BYTES support?

Reply:
I had initially planned to do it later. But I have added support now looking at what selection operator does